### ISSUE TITLE
Add ability to run operations in parallel

### ISSUE THREAD

**Issue by smaslennikov** [minio/mc#3450]
Apologies, this is a feature request rather than a bug report. If there's a better place for it, I'll happily move it.

I'm aware of the [multipart transfer functionality](https://github.com/minio/mc/issues/1974), but this request is more to imitate `gsutil -m`: "Causes supported operations (acl ch, acl set, cp, mv, rm, rsync, and setmeta) to run in parallel. This can significantly improve performance if you are performing operations on a large number of files over a reasonably fast network connection."

**Comment by harshavardhana** [comment]
mc already uses multiple operations in parallel automatically depending on the egress bandwidth and tapers out automatically as well. 

Not sure what is needed from `mc` here.

**Comment by smaslennikov** [comment]
@harshavardhana I haven't looked much under the hood of `mc` and don't know what parallelism is used for specifically. In `gsutil -m`, many small files are transferred in parallel - if that's automatically done with bare `mc`, that works for me.

Thanks!

**Comment by harshavardhana** [comment]
> @harshavardhana I haven't looked much under the hood of `mc` and don't know what parallelism is used for there specifically. In `gsutil -m`, many small files are transferred in parallel - if that's automatically done with bare `mc`, that works for me.

This is automatically done in `mc`

**Comment by AlexisEspinosaGayosso** [comment]
Dear @harshavardhana,
How are these parallel transfers controlled by the available resources?

(I see above in your responses that the parallel transfer is performed automatically, so I guess it checks for the available resources and perform the transfer in parallel if resources are available. And I also see that users do not have control over the parallel parameters on mc.)

I'm asking you this because I do not see any difference in elapsed transfer time of a test directory of 20 files (1.5Gb each) to our ceph/S3 repository. I mean difference when allocating 4 cores to the job or when allocating just a single core to the job. (Our job scheduler is slurm)

So, in other words, should I always allocate a single core to mc jobs? Or, is there a way to enhance parallel capabilities of mc by allocating more than 1 core to the job?

Many thanks.

**Comment by harshavardhana** [comment]
> Dear @harshavardhana, How are these parallel transfers controlled by the available resources?
> 
> (I see above in your responses that the parallel transfer is performed automatically, so I guess it checks for the available resources and perform the transfer in parallel if resources are available. And I also see that users do not have control over the parallel parameters on mc.)
> 
> I'm asking you this because I do not see any difference in elapsed transfer time of a test directory of 20 files (1.5Gb each) to our ceph/S3 repository. I mean difference when allocating 4 cores to the job or when allocating just a single core to the job. (Our job scheduler is slurm)
> 
> So, in other words, should I always allocate a single core to mc jobs? Or, is there a way to enhance parallel capabilities of mc by allocating more than 1 core to the job?
> 
> Many thanks.

It's GOMAXPROCS
